File: reverse-linked-list/reverse-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rev(ListNode* node) {
        if(node->next!=nullptr){
            ListNode* nxt = node->next;
            node->next = nullptr;
            ListNode* head = rev(nxt);
            nxt->next = node;
            return head;
        }
        return node;
    }
    ListNode* reverseList(ListNode* head) {
        if(!head) return head;
        return rev(head);
    }
};

File: reverse-linked-list/reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        node, rev = head, None
        while node:
            next, node.next = node.next, rev
            rev, node= node, next
        return rev